[otap-df-pdata]Add zero-copy view for OTAP logs format

lalitb opened the pull request:
Fixes: #1441 

Implements `OtapLogsView` - a zero-copy view over Arrow RecordBatch data for OTAP logs.

## Use Cases
This view is useful for  exporters that need to iterate over the OTAP log data:

- Geneva Exporter: Enables the exporter to iterate directly over Arrow data to generate custom payloads, avoiding the costly round-trip of converting to OTLP bytes and decoding them back to objects.
- Log Analytics: Allows efficient, zero-copy access to log fields for serialization to JSON.

- General Data Access: Provides a standard, efficient way for any component to read the hierarchical OTLP structure (Resource → Scope → Log) without needing to understand the underlying Arrow complexity.

Without views: OTAP → OTLP bytes → decode → iterate (slower) 
With views: OTAP → view → iterate (faster, zero-copy)

 ## Design

- Stores column references and accessors (fixed overhead)
- Pre-computes indices for resource/scope/log hierarchy
- Actual telemetry data remains in Arrow RecordBatch (no copying)

## Benchmark Results

  System: Apple M4 Pro, 12 cores, 24 GB RAM

  | Logs | Zero-Copy View | OTAP→OTLP→Process |
  |------|----------------|-------------------|
  |   10 |      1.48 µs   |      6.43 µs      |
  |  100 |      9.86 µs   |     40.40 µs      |
  | 1000 |    121.42 µs   |    383.11 µs      |

  OTAP→OTLP→Process = full cost without views (OTAP→OTLP conversion + OTLP decode + iterate)

  ## Changes

 - Added `OtapLogsView` with hierarchical iteration (Resource → Scope → LogRecord → Attributes)
 - Added column caching to avoid repeated schema lookups
 - Updated Geneva exporter:
    - Added example view-based code (commented out) to demonstrate intended usage
    - Currently uses OTAP→OTLP fallback conversion until geneva-uploader 0.4.0+ supports view API
    - All pipelines (OTLP bytes and OTAP Arrow) continue to work

jmacd reviewed (approved):
Excellent!

I have one thought, non-blocking feedback looking for additional ways to test this.

I was thinking that since the former releases of geneva_uploader were based on OTLP protocol message objects, that I'd see OTLP test message objects used as input and corresponding test expectations written for the Geneva output we expect.

Possibly, when the geneva_uploader component releases with view support, we could see a few tests with input OlpProtoMessages and output Geneva structs added, so that we're testing an end-to-end without hand-crafted OtapArrowRecords.

lquerel commented on rust/otap-dataflow/crates/pdata/src/views/otap/logs.rs:
```diff
@@ -0,0 +1,2004 @@
+    None
+}
+
+#[cfg(test)]
```
Did you test scenarios with log batches that contain no resource attributes or fields, or no scope attributes/fields?

lquerel reviewed (commented):
Overall looks good, made few comments.

I’m wondering whether all cases have been tested, especially those where no attributes are present.

@albertlockett could take a look at this PR.

lalitb commented:
> Possibly, when the geneva_uploader component releases with view support, we could see a few tests with input OlpProtoMessages and output Geneva structs added, so that we're testing an end-to-end without hand-crafted OtapArrowRecords.

Good suggestion!  I'll look into adding OTLP → Geneva struct tests that validate the encoding without needing backend connectivity. That would give us better test coverage of the full conversion path.

albertlockett commented on rust/otap-dataflow/crates/pdata/src/views/otap/logs.rs:
```diff
@@ -0,0 +1,2050 @@
+}
+
+/// Iterator over attributes for a resource
+pub struct OtapResourceAttributeIter<'a> {
```
It seems like `OtapResourceAttributeIter`, `OtapScopeAttributeIter` and `OtapAttributeIter` all have a similar implementation. Is there any reason we couldn't combine these into one type?

lalitb commented:
> I had a few minor comments about where we could reuse some code to avoid duplication. I'm happy to approve as-is if these suggestions aren't sensible so let me know what you think :)

  @albertlockett - All valid points, thanks! I've addressed the code duplication you raised. Additionally, I've improved the attribute iterator handling, added dictionary encoding support for ID columns, and expanded test coverage.

 I would be doing similar implementation for traces once this is merged.

albertlockett reviewed (approved):
approved! Thanks again @lalitb for the contribution. @lquerel good to merge it
